File: app/src/main/java/com/haishinkit/rtmp/message/RTMPSetPeerBandwidthMessage.java
package com.haishinkit.rtmp.message;

import com.haishinkit.lang.IRawValue;
import com.haishinkit.rtmp.RTMPConnection;
import com.haishinkit.rtmp.RTMPSocket;

import java.nio.ByteBuffer;

public final class RTMPSetPeerBandwidthMessage extends RTMPMessage {
    private static int CAPACITY = 5;

    public enum Limit implements IRawValue<Byte> {
        HARD((byte) 0x00),
        SOFT((byte) 0x01),
        DYNAMIC((byte) 0x10),
        UNKNOWN((byte) 0x00);

        public static Limit rawValue(final byte rawValue) {
            switch (rawValue) {
                case 0x00:
                    return HARD;
                case 0x01:
                    return SOFT;
                case 0x02:
                    return DYNAMIC;
                default:
                    return UNKNOWN;
            }
        }

        private final byte rawValue;

        Limit(final byte rawValue) {
            this.rawValue = rawValue;
        }

        public Byte rawValue() {
            return rawValue;
        }
    }

    private int size = 0;
    private Limit limit = Limit.HARD;

    public RTMPSetPeerBandwidthMessage() {
        super(Type.BANDWIDTH);
    }

    public int getSize() {
        return size;
    }

    public RTMPSetPeerBandwidthMessage setSize(final int size) {
        this.size = size;
        return this;
    }

    public Limit getLimit() {
        return limit;
    }

    public RTMPSetPeerBandwidthMessage setLimit(final Limit limit) {
        this.limit = limit;
        return this;
    }

    @Override
    public ByteBuffer encode(final RTMPSocket socket) {
        ByteBuffer buffer = ByteBuffer.allocate(CAPACITY);
        buffer.putInt(size);
        buffer.put(limit.rawValue());
        return buffer;
    }

    @Override
    public RTMPMessage decode(final RTMPSocket socket, final ByteBuffer buffer) {
        return setSize(buffer.getInt()).setLimit(Limit.rawValue(buffer.get()));
    }

    @Override
    public RTMPMessage execute(final RTMPConnection connection) {
        connection.getSocket().setBandwidth(getSize());
        return this;
    }
}

File: haishinkit/src/main/java/com/haishinkit/rtmp/messages/RTMPSetPeerBandwidthMessage.kt
package com.haishinkit.rtmp.messages

import com.haishinkit.rtmp.RTMPConnection
import com.haishinkit.rtmp.RTMPSocket
import java.nio.ByteBuffer

/**
 * 5.4.5. Set Peer Bandwidth (6)
 */
internal class RTMPSetPeerBandwidthMessage : RTMPMessage(RTMPMessage.Type.BANDWIDTH) {
    enum class Limit(val rawValue: Byte) {
        HARD(0x00),
        SOFT(0x01),
        DYNAMIC(0x02),
        UNKNOWN(Byte.MAX_VALUE);
    }

    var size = 0
        private set
    var limit = Limit.HARD
        private set

    override fun encode(socket: RTMPSocket): ByteBuffer {
        val buffer = ByteBuffer.allocate(CAPACITY)
        buffer.putInt(size)
        buffer.put(limit.rawValue)
        return buffer
    }

    override fun decode(buffer: ByteBuffer): RTMPMessage {
        size = buffer.int
        var limit = buffer.get()
        this.limit = Limit.values().first { n -> n.rawValue == limit }
        return this
    }

    override fun execute(connection: RTMPConnection): RTMPMessage {
        connection.socket.bandWidth = size
        return this
    }

    companion object {
        private const val CAPACITY = 5
    }
}
